_posts/2022-09-15-redis-job-queue.adoc

```diff
@@ -0,0 +1,302 @@
+---
+layout: post
+title: 'How to implement a job queue with Redis'
+date: 2022-09-15
+tags: redis
+synopsis: Implementing a job queue using Redis list and the Quarkus Redis extension
+author: cescoffier
+---
+
+In https://quarkus.io/blog/redis-api-intro/[how to cache with Redis], we implemented a simple cache backed by Redis. +
+That's just one use case of Redis.
+Redis is also used as a messaging server to implement the processing of background jobs or other kinds of messaging tasks.
+This post explores implementing this pattern with Quarkus and the new Redis data source API.
+
+== Job Queues and Supes!
+
+A job queue is a data structure storing _execution_ requests.
+Job dispatchers submit the tasks they want to execute in that data structure.
+On the other side, job consumers poll the requests and execute them.
+
+image::/assets/images/posts/redis-job-queue/pattern.png[align="center"]
+
+There are plenty of variants of that pattern, so let's focus on the following application.
+We have an application managing heroes and villains.
+The application offers the possibility to simulate a fight between a random hero and a random villain.
+The fight simulation is delegated to _fight simulators_, applications dedicated to that task.
+
+image::/assets/images/posts/redis-job-queue/application.png[align="center"]
+
+In this context, the main application submits the fight request to the job queue.
+Then, the fight simulators poll the submitted fight request and execute them.
+
+The fight outcomes are communicated using another Redis feature: pub/sub communication.
+The simulators send the outcome to a _channel_ consumed by the application.
+The application then broadcasts these outcomes to a web page.
+
+This post only discusses the interaction with Redis.
+The rest of the application is straightforward and just uses RESTEasy Reactive and Hibernate ORM with Panache.
+You can find the full code of the application on https://github.com/cescoffier/quarkus-redis-job-queue-demo.
+
+== Submitting jobs
+
+The first task is to modelize the job queue.
+We are using a https://redis.io/docs/data-types/#lists[Redis list] to store the https://github.com/cescoffier/quarkus-redis-job-queue-demo/blob/main/fight-simulator/src/main/java/me/escoffier/quarkus/redis/fight/FightRequest.java[FightRequest].
+
+[source,java]
+----
+package me.escoffier.quarkus.redis.fight;
+
+public record FightRequest(String id, Hero hero, Villain villain) {
+
+}
+----
+
+Redis lists distinguish the left side of the list from the right side of the list.
+This distinction allows implementing a FIFO queue where we write on the left side and consume from the right side.
```

Comment:
Maybe instead of `left` and `right` we should use `beginning` and `end`? I find the former rather confusing.

_EDIT:_ I guess you are using these terms because that is what Redis uses - in that case, I would say you made the correct decision

Reply:
Yes, Redis uses left and right when dealing with redis list.
`BRPOP` stands for _blocking right pop_ .... We can also debate about the blocking aspect ;-)